### ISSUE TITLE
optimization breaks variable length array parameters

### ISSUE THREAD

**Issue by 1-Olivier** [imageworks/OpenShadingLanguage#655]
If several layers in a network are instances of the same shader and that shader has a variable length array parameter, trying to set a different value for the two instances will fail. The same value will be used by both. This does not happen at O0 osl optimization level. This archive [simple_surface.tar.gz](https://github.com/imageworks/OpenShadingLanguage/files/350164/simple_surface.tar.gz) is a fairly minimal reproduction case:
```
> testshade --group shadergroup
0.000000 0.000000 2.000000

> testshade -O0 --group shadergroup
1.000000 0.000000 1.000000
```

**Comment by lgritz** [comment]
I see what's happening! The "instance merging" is being confused and not seeing that the different values passed to the varying length array is not the same, and incorrectly merging the two shaders.

As a workaround, you can disable this by setting the ShadingSystem attribute "opt_merge_instances" to zero.  We can test this:

    OSL_OPTIONS=opt_merge_instances=0 testshade --group shadergroup

and yes, now it gives the correct answer, without needing to disable any other optimizations. (-O0 is useless for production)

Let me see if I can get this fixed this afternoon. Stay tuned.

**Comment by 1-Olivier** [comment]
Thanks, that indeed fixes it. I think we can live without instance merging for a few days or weeks without anyone crying over it.

**Comment by lgritz** [comment]
Fix proposed in #656.

After approval and some testing, I'll also backport this into the release branch.

**Comment by lgritz** [comment]
I've merged the fix into both master and RB-1.7.
